torch/nn/_functions/thnn/auto.py

```diff
@@ -86,41 +108,47 @@ def has_argument(fn, name):
         buffers['acc_grad_parameters'] = _find_buffers(
             acc_grad_parameters.arguments[3:], ignored_args)
 
-    # This and __init__ assume that only the last argument can be
+    # This assumes that only the last argument can be
     # an inplace flag
     is_inplace = update_output.arguments[-1].name == 'inplace'
 
-    def __init__(self, *args):
-        if is_inplace:
-            InplaceFunction.__init__(self, args[-1])
-        else:
-            Function.__init__(self)
-        self.additional_args = list(args)
-
-    def _initialize_buffers(self, fn_name):
-        additional_args = self.additional_args
+    @staticmethod
+    def _initialize_buffers(ctx, fn_name):
+        additional_args = ctx.additional_args
         for idx, name in buffers[fn_name]:
             # TODO: some buffers are necessary only for update output and can be
             # freed right afterwards
-            buffer = self.buffers[name]
+            buffer = ctx.buffers[name]
             additional_args = additional_args[:idx] + [buffer] + additional_args[idx:]
         return tuple(additional_args)
 
-    def forward(self, input, *params):
-        self._backend = type2backend[type(input)]
+    @staticmethod
+    def forward(ctx, input, *params):
+        ctx._backend = type2backend[type(input)]
 
+        ctx.additional_args = []
+        tensor_param_list = []
         for param in params:
-            if type(param) != type(input):
-                raise RuntimeError("input type ({}) doesn't match the type of "
-                                   "a parameter tensor ({})".format(torch.typename(input),
-                                                                    torch.typename(param)))
+            if torch.is_tensor(param):
+                if type(param) != type(input):
+                    raise RuntimeError("input type ({}) doesn't match the type of "
+                                       "a parameter tensor ({})".format(torch.typename(input),
+                                                                        torch.typename(param)))
+                tensor_param_list.append(param)
+            else:
+                ctx.additional_args.append(param)
 
+        tensor_params = tuple(tensor_param_list)
+        if is_inplace:
+            ctx.inplace = params[-1]
         # Allocate temporary buffers and insert them into additional_args
-        self.buffers = defaultdict(type(input))
-        additional_args = self._initialize_buffers('update_output')
+        ctx.buffers = defaultdict(type(input))
+        # we need to call __func__ here because the method is static; we could move it
+        # to outside the class but that would take some restructuring.
+        additional_args = _initialize_buffers.__func__(ctx, 'update_output')
```

Comment:
You could just capture `_initialize_buffer` as part of the closure. There's no need to assign it as a class method and use these hacks